Fixes #4421 - HttpClient support for PROXY protocol.

sbordet opened the pull request:
Implemented support for the PROXY protocol in HttpClient.

Introduced Request.tag(Object) to tag requests that belong
to the same group (e.g. a client address) so that they can
generate a different destination.

The tag object may implement ClientConnectionFactory.Decorator
so that it can decorate the HttpDestination ClientConnectionFactory
and therefore work both with and without forward proxy configuration.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

sbordet committed:
Fixes #4421 - HttpClient support for PROXY protocol.

Implemented support for the PROXY protocol in HttpClient.

Introduced Request.tag(Object) to tag requests that belong
to the same group (e.g. a client address) so that they can
generate a different destination.

The tag object may implement ClientConnectionFactory.Decorator
so that it can decorate the HttpDestination ClientConnectionFactory
and therefore work both with and without forward proxy configuration.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/Origin.java:
```diff
@@ -63,14 +80,23 @@ public boolean equals(Object obj)
     {
-        int result = scheme.hashCode();
-        result = 31 * result + address.hashCode();
+        return Objects.hash(scheme, address, tag);
```
Why no `equals` implementation if you have a `hashCode` impl?

sbordet replied:
Implementation of `equals()` is just above this method.

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/Origin.java:
```diff
@@ -48,6 +60,11 @@ public Address getAddress()
         return address;
     }
 
+    public Object getTag()
```
Is javadoc needed here to indicate that this is the same tag from the associated request?

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java:
```diff
@@ -0,0 +1,489 @@
+            private int srcPort;
+            private String dstIP;
+            private int dstPort;
+            private Map<Integer, byte[]> vectors;
```
Not enough javadoc on this class.... what is a vector?

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java:
```diff
@@ -0,0 +1,489 @@
+            return new ProxyProtocolConnectionV2(endPoint, executor, getClientConnectionFactory(), context, info);
+        }
+
+        public static class Info implements ClientConnectionFactory.Decorator
```
Shouldn't the V2 Info class extend the V1 Info class?  Or better yet just have a single Info class that can be used for both and that has a better name than `Info`.

gregw commented on jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientProxyProtocolTest.java:
```diff
@@ -0,0 +1,240 @@
+        int serverPort = connector.getLocalPort();
+
+        int clientPort = ThreadLocalRandom.current().nextInt(1024, 65536);
+        V1.Info info = new V1.Info("TCP4", "127.0.0.1", clientPort, "127.0.0.1", serverPort);
```
This example makes no sense to me?   The user should not be dealing with IP addresses and ports like this.  It is not entirely clean which client this is talking about -  specifically it looks like I need to know my own client port before it has even been allocated (which it is not as I guess it is the port that we are forwarding??).
Why must the user specify information that is in the request anyway, such as the server and serverPort?  Does it have to be server IP or can it be server name?  What about servers with multiple IPs and DNS balancing?

gregw reviewed (changes requested):
Don't like this at all!

The Tag mechanism appears loosely typed but will fail with a class cast exception if a tag of some random type is assigned.

The Info classes are poorly named and appear to have duplicate information in them (the server IP/port).  It is not clear in their usage that they are describing a proxy tunnel.

I think the language of the API needs to be strongly typed and explicitly about Proxy, Tunnel, Connection, ForwardFor or something better than Tag and Info, which could be anything.

Does this need to be linked somehow with setting the Forwarded headers?

If this design is indeed the only viable solution, then it needs a lot more explanation what the Info and Tag are communicating and how they should be used.

gregw commented:
@sbordet please remember to link #4421 in the PR description!

sbordet committed:
Fixes #4421 - HttpClient support for PROXY protocol.

Updates after review.
Renamed Info -> Tag.
Added convenience constructors to class Tag.
Added test and server support for PROXY V1 UNKNOWN.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>

joakime commented:
> Don't like this at all!
> 
> The Tag mechanism appears loosely typed but will fail with a class cast exception if a tag of some random type is assigned.

Personally, I like the tag.

It's intentionally loosely typed, and can apply to anything related to destination resolution.
Even things that have nothing to do with proxies.

I could use it to tag requests with a common object that my application knows and uses.

I could use the tag to section off destinations by something arbitrary (like load clients and a minimum physical connections config).

Just because the tag can *also* apply to a proxy is a curiosity and isn't in conflict with the tag idea.
The information contained in the tag can be extended, to cover all these bases.
And the tag can also be used to satisfy highly custom connection ideas that even fall outside of the proxy concepts.

gregw commented:
@joakime during the review, the name 'Tag' grew on me, specially with `Info` renamed to `Tag` so it is clear that it a class that can be used to tag requests and/or origins for proxy protocol.

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java:
```diff
@@ -0,0 +1,561 @@
+                this(null, 0);
+            }
+
+            public Tag(String srcIP, int srcPort)
```
probably should javadoc these constructors to say that null values mean that the proxy protocol will take the value from the EndPoint when then tag is applied to a specific connection (and not defaulted at construction).

gregw replied:
Ditto v1

gregw commented on jetty-client/src/main/java/org/eclipse/jetty/client/ProxyProtocolClientConnectionFactory.java:
```diff
@@ -0,0 +1,561 @@
+        {
+            try
+            {
+                InetSocketAddress localAddress = endPoint.getLocalAddress();
```
Do all our endpoints return non null InetAddresses?   What if this is applied to a UDP endpoint?

gregw reviewed (approved):
It needs a little bit more javadoc in places.... and perhaps a null check for strange endpoints... but LGTM and can be merged.

sbordet committed:
Fixes #4421 - HttpClient support for PROXY protocol.

Updates after review.
Improved javadocs.

Signed-off-by: Simone Bordet <simone.bordet@gmail.com>
